### PR TITLE
[TS migration] Migrate 'getPlaidLinkTokenParameters' lib to TypeScript

### PR DESCRIPTION
**Pull request by VickyStash**
### Details

Migrate 'getPlaidLinkTokenParameters' lib to TypeScript

### Fixed Issues

$ https://github.com/Expensify/App/issues/24858
PROPOSAL: N/A

### Tests

- [x] Verify that no errors appear in the JS console

### Offline tests

N/A

### QA Steps

- [x] Verify that no errors appear in the JS console

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android / native
    - [x] Android / Chrome
    - [x] iOS / native
    - [x] iOS / Safari
    - [x] MacOS / Chrome / Safari
    - [x] MacOS / Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] If we are not using the full Onyx data that we loaded, I've added the proper selector in order to ensure the component only re-renders when the data it is using changes
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Web</summary>

https://github.com/Expensify/App/assets/23176449/f6fdcf24-5bf4-4587-a45f-3d8353b8dc93

</details>

<details>
<summary>Mobile Web - Chrome</summary>

https://github.com/Expensify/App/assets/23176449/ebcf1ea3-e6fd-4ab7-bd26-44d6a256adc4

</details>

<details>
<summary>Mobile Web - Safari</summary>

https://github.com/Expensify/App/assets/23176449/31a7752d-3a8f-4a50-b103-1fcb06320625

</details>

<details>
<summary>Desktop</summary>

https://github.com/Expensify/App/assets/23176449/dc83bd9d-4246-49ea-9aec-8964dc6c4c24

</details>

<details>
<summary>iOS</summary>

https://github.com/Expensify/App/assets/23176449/c82ad092-ace9-419d-9c4e-aca17ffa9ded

</details>

<details>
<summary>Android</summary>

https://github.com/Expensify/App/assets/23176449/a65b4ef8-55a3-4627-be91-ddec45d5035f

</details>

### CODE REVIEW

**Comment by VickyStash** [comment]
I have read the CLA Document and I hereby sign the CLA

File: src/libs/getPlaidLinkTokenParameters/types.ts [hunk 1]
```diff
@@ -0,0 +1,7 @@
+type PlaidLinkTokenParameters =
+    | { android_package: string } // eslint-disable-line @typescript-eslint/naming-convention
+    | { redirect_uri: string }; // eslint-disable-line @typescript-eslint/naming-convention
```
Comment (fabioh8010) [inline src/libs/getPlaidLinkTokenParameters/types.ts:3]:
```suggestion
type PlaidLinkTokenParameters = {
    // eslint-disable-next-line @typescript-eslint/naming-convention
    android_package?: string;

    // eslint-disable-next-line @typescript-eslint/naming-convention
    redirect_uri?: string;

    allowDebit?: boolean;
    bankAccountID?: number;
};
```

Since we have a platform file which uses `android_package` and others which uses `redirect_uri`, a better representation of this data would be with these parameters as optional ones.

Also, `getPlaidLinkTokenParameters` is only used in `src/libs/actions/Plaid.js` and it seems that they add `allowDebit` and `bankAccountID` to this params, so we should include them as well.

File: src/libs/getPlaidLinkTokenParameters/index.ios.ts [hunk 2]
```diff
@@ -0,0 +1,8 @@
+import CONFIG from '../../CONFIG';
+import GetPlaidLinkTokenParameters from './types';
+
+const getPlaidLinkTokenParameters: GetPlaidLinkTokenParameters = () => ({
+    redirect_uri: `${CONFIG.EXPENSIFY.NEW_EXPENSIFY_URL}partners/plaid/oauth_ios` // eslint-disable-line @typescript-eslint/naming-convention
```
Comment (fabioh8010) [inline src/libs/getPlaidLinkTokenParameters/index.ios.ts:5]:
```suggestion
    // eslint-disable-line @typescript-eslint/naming-convention
    redirect_uri: `${CONFIG.EXPENSIFY.NEW_EXPENSIFY_URL}partners/plaid/oauth_ios`
```

nitpick

File: src/libs/getPlaidLinkTokenParameters/index.android.ts [hunk 3]
```diff
@@ -0,0 +1,8 @@
+import CONST from '../../CONST';
+import GetPlaidLinkTokenParameters from './types';
+
+const getPlaidLinkTokenParameters: GetPlaidLinkTokenParameters = () => ({
+    android_package: CONST.ANDROID_PACKAGE_NAME, // eslint-disable-line @typescript-eslint/naming-convention
```
Comment (fabioh8010) [inline src/libs/getPlaidLinkTokenParameters/index.android.ts:5]:
```suggestion
    // eslint-disable-line @typescript-eslint/naming-convention
    android_package: CONST.ANDROID_PACKAGE_NAME,
```

nitpick

**Review by fabioh8010 — APPROVED** [review]

File: src/libs/getPlaidLinkTokenParameters/types.ts [hunk 4]
```diff
@@ -0,0 +1,14 @@
+type PlaidLinkTokenParameters = {
+    // eslint-disable-next-line @typescript-eslint/naming-convention
+    android_package?: string;
+
+    // eslint-disable-next-line @typescript-eslint/naming-convention
+    redirect_uri?: string;
+
+    allowDebit?: boolean;
+    bankAccountID?: number;
```
Comment (rezkiy37) [inline src/libs/getPlaidLinkTokenParameters/types.ts:9]: Should we add captions for these properties?

**Review by rezkiy37 — COMMENTED** [review]
LGTM

File: src/libs/getPlaidLinkTokenParameters/types.ts (hunk 4 above)
Comment (VickyStash) [reply src/libs/getPlaidLinkTokenParameters/types.ts:9]: Since we don't have a requirement to caption TS types, and this type is not widely used, I think we can leave it as is.

**Review by aimane-chnaif — APPROVED** [review]

**Review by blazejkustra — APPROVED** [review]
LGTM, SWM still need to do the checklist.

**Comment by BartoszGrajdek** [comment]
## Reviewer Checklist

- [x] I have verified the author checklist is complete (all boxes are checked off).
- [x] I verified the correct issue is linked in the `### Fixed Issues` section above
- [x] I verified testing steps are clear and they cover the changes made in this PR
    - [x] I verified the steps for local testing are in the `Tests` section
    - [x] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [x] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [x] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I verified tests pass on **all platforms** & I tested again on:
    - [x] Android / native
    - [x] Android / Chrome
    - [x] iOS / native
    - [x] iOS / Safari
    - [x] MacOS / Chrome / Safari
    - [x] MacOS / Desktop
- [x] If there are any errors in the console that are unrelated to this PR, I either fixed them (preferred) or linked to where I reported them in Slack
- [x] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Web</summary>



https://github.com/Expensify/App/assets/41169160/d1404ae9-2de1-4961-93ce-ba3a40aaa826



</details>

<details>
<summary>Mobile Web - Chrome</summary>



https://github.com/Expensify/App/assets/41169160/bb203f70-2378-4186-8d1a-de5e00aeb7ce



</details>

<details>
<summary>Mobile Web - Safari</summary>



https://github.com/Expensify/App/assets/41169160/4ceab729-1fa0-4087-aa34-1eae7994799c



</details>

<details>
<summary>Desktop</summary>



https://github.com/Expensify/App/assets/41169160/3193ad2e-df94-40e0-a2a2-2d143376d2ed



</details>

<details>
<summary>iOS</summary>



https://github.com/Expensify/App/assets/41169160/a97d3352-18fe-4484-8b7b-0358939ae151



</details>

<details>
<summary>Android</summary>



https://github.com/Expensify/App/assets/41169160/743e0291-8e3c-426e-a3e7-5bb9a0584aa0



</details>

File: src/libs/getPlaidLinkTokenParameters/index.ts [hunk 5]
```diff
@@ -1,7 +1,12 @@
 import ROUTES from '../../ROUTES';
 import CONFIG from '../../CONFIG';
+import GetPlaidLinkTokenParameters from './types';
 
-export default () => {
+const getPlaidLinkTokenParameters: GetPlaidLinkTokenParameters = ()=> {
```
Comment (cead22) [inline src/libs/getPlaidLinkTokenParameters/index.ts:5]:
```suggestion
const getPlaidLinkTokenParameters: GetPlaidLinkTokenParameters = () => {
```

File: src/libs/getPlaidLinkTokenParameters/types.ts [hunk 6]
```diff
@@ -0,0 +1,14 @@
+type PlaidLinkTokenParameters = {
+    // eslint-disable-next-line @typescript-eslint/naming-convention
```
Comment (cead22) [inline src/libs/getPlaidLinkTokenParameters/types.ts:2]: NAB but why are we suppressing this several times in this PR? Would it make sense to abide by this convention as part of this PR or would that require a lot of updated that are out of scope for this PR?

**Review by cead22 — CHANGES_REQUESTED** [review]

File: src/libs/getPlaidLinkTokenParameters/types.ts (hunk 6 above)
Comment (VickyStash) [reply src/libs/getPlaidLinkTokenParameters/types.ts:2]: @cead22 This convention is more of an exception for this specific case, so I think we shouldn't apply it everywhere.

Comment (cead22) [reply src/libs/getPlaidLinkTokenParameters/types.ts:2]: Can you help me understand this a little better? I don't think it's truly an exception since there's 5 occurrences in this PR, which is very small.

What are the reasons for not updating the variable names to meet the naming convention?

Comment (VickyStash) [reply src/libs/getPlaidLinkTokenParameters/types.ts:2]: @cead22 These variables are parameters that are later sent to the API
[API.read('OpenPlaidBankLogin', params, {optimisticData});](https://github.com/Expensify/App/blob/main/src/libs/actions/Plaid.js#L36C5-L36C5)

So even if we rename it here, then before sending it to the API, we will need to do something like `{redirect_uri: redirectUri}`, and it also can cause a convention warning

Comment (cead22) [reply src/libs/getPlaidLinkTokenParameters/types.ts:2]: That makes sense, so I think the correct way to complete this issue is
- Update the backend to accept redirectURI and redirect_uri (cc @nkuoch)
- Update JS to use the right naming convention

If we don't do this, then we'll continue breaking our conventions forever, and continue adding exceptions, which defeats the purpose of having them.

We need to think about our code holistically and continuously update the code surrounding the code we edit

Comment (blazejkustra) [reply src/libs/getPlaidLinkTokenParameters/types.ts:2]: I agree with @cead22, however I don't think it's in the scope for this PR. 

Once majority of files are migrated to typescript we can search for these occurrences and fix these names more easily - TS will warn us if names are migrated and used correctly. What do you think? @cead22 @VickyStash

Comment (tgolen) [reply src/libs/getPlaidLinkTokenParameters/types.ts:2]: I agree with @cead22 and I would avoid pushing this off into another PR. I did something similar with `filename` and `receipeFilname` recently. 

I understand it's not strictly in the scope of this PR, but If we wrapped every PR with a second PR that was basically "clean up all the code around the previous PR X" I think that would get out of hand.

In the interest of always leaving code in a better state than it was found, I think it should be cleaned up here.

If it's a matter that ends up significantly increase your workload, we are pretty fair and can increase compensation using our discretion, so I wouldn't worry too much about that (if it was at all part of your thinking).

Comment (fabioh8010) [reply src/libs/getPlaidLinkTokenParameters/types.ts:2]: @tgolen @cead22 So, let's hold this PR and create another GH issue to change those parameters in the backend. Once that issue is done we can go back to this one and finish it. Sounds good? Do you want me to create the GH issue for you?

Comment (cead22) [reply src/libs/getPlaidLinkTokenParameters/types.ts:2]: I made the [GH issue](https://github.com/Expensify/Expensify/issues/318749) (in E/Expensify). I can post here when the change to accept both redirect_uri and redirectURI is live

Comment (fabioh8010) [reply src/libs/getPlaidLinkTokenParameters/types.ts:2]: Thanks @cead22 🙌

Comment (cead22) [reply src/libs/getPlaidLinkTokenParameters/types.ts:2]: PR is merged. It should be live on Tuesday

Comment (VickyStash) [reply src/libs/getPlaidLinkTokenParameters/types.ts:2]: @cead22 Got it, thank you!

Comment (VickyStash) [reply src/libs/getPlaidLinkTokenParameters/types.ts:2]: @cead22 Just want to confirm, have you also updated `android_package` field to be accepted as both `android_package` and `androidPackage`?

Comment (cead22) [reply src/libs/getPlaidLinkTokenParameters/types.ts:2]: 🤦 I did not, I'll do that now

Comment (cead22) [reply src/libs/getPlaidLinkTokenParameters/types.ts:2]: Yes, they were deployed to production yesterday!

Comment (VickyStash) [reply src/libs/getPlaidLinkTokenParameters/types.ts]: @cead22 Thank you! I've updated the PR

**Review by cead22 — APPROVED** [review]

**Comment by cead22** [comment]
Thank you @VickyStash for the cleanup on the var names
